### DISCUSSION TITLE
PDF exporter

### DISCUSSION THREAD

**Discussion by bdhrs** [digitalpalidictionary/dpd-db#15] [Ideas]
The suggestion came up to render a PDF of DPD, which can be used in a number of scenarios that software cannot. Where to start the process? @gambhiro I know your PDF's are always immaculately styled, do you have a template that can be adapted? @Devamitta

**Comment by Devamitta** [comment]
I think can be something similar to [combined-dictionary](https://github.com/simsapa/simsapa-dictionary/releases) format.

![image](https://github.com/digitalpalidictionary/dpd-db/assets/39419221/3bb6d72f-742a-4aef-bfe4-1b65b652630b)

**Comment by bergentroll** [comment]
May I say there are two nice solutions to make PDF with Python:
- [ReportLab](https://docs.reportlab.com/)
- [PyLaTeX](https://github.com/JelteF/PyLaTeX)

The second depends on large LaTeX runtime, but may give more "book" style.

**Comment by gambhiro** [comment]
I use LaTeX to make books and it's certainly suitable for the task, but getting the style macros right is arcane, compiling the PDF is slow, and there is no incremental build. (One may not need a complete TeXLive setup, the [Tectonic](https://github.com/tectonic-typesetting/tectonic) project aims to create a minimal install.)

I recommend this project instead: https://typst.app/

They aim to be a LaTeX replacement but with modern programming. It has a decent template and scripting function. It requires a single binary download to produce a PDF. It can watch a file for changes and make incremental PDF compilation.

You could approach it by writing templates like the HTML rendering, but producing the Typst syntax instead.

Another (more technically challenging) way would be to read CSV or JSON with Typst scripts, maybe a plugin with Rust, and render the words from Typst.

This is the table syntax, they try to impress you with the math but it's just a comma separated list of items: https://typst.app/docs/reference/layout/table/

Core templates: https://github.com/typst/templates

Community projects: https://github.com/qjcg/awesome-typst

HTML and EPUB rendering is on their roadmap:

https://typst.app/docs/roadmap/

**Comment by gambhiro** [comment]
Pandoc now has [both reader](https://github.com/jgm/pandoc/issues/8740) and writer support for typst.

So if you generate typst format, typst can build the PDF and pandoc can produce HTML and EPUB.

It's a useful bridge until typst implements those exports.

**Comment by Devamitta** [comment]
Completed with https://github.com/digitalpalidictionary/dpd-db/tree/main/exporter/pdf
